Add Gaudi support to benchmarks/dynamo

kfojcik-intel opened the pull request:
This is a reopen of PR [#122960](https://github.com/pytorch/pytorch/pull/122960) which due to inactivity was closed. 
The change enables benchmarks/dynamo/* with Gaudi.
It is similar to that proposed in #122960, we only removed the part of the GPU Migration Toolkit import that is no longer needed in the primary scripts.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

kfojcik-intel committed:
Add Gaudi support to benchmarks/dynamo

desertfire commented on benchmarks/dynamo/common.py:
```diff
@@ -991,7 +991,7 @@ def maybe_mark_profile(*args, **kwargs):
             )
         else:
-            frozen_model_iter_fn = torch._dynamo.run(model_iter_fn)
+            frozen_model_iter_fn = optimize_ctx(model_iter_fn)
```
This looks fine, but can you elaborate more on why you need to make the change?

kfojcik-intel replied:
Hi @desertfire, 
Sorry for late response on that. In general, `speedup_experiment` calculates speedup over eager. Imo, it should take defined optimizations in account like e.g. compile backend. This will not happen here if we do not take `optimize_ctx`. It is made similarly for other performance metrics, e.g. `dynamo_latency, dynamo_peak_mem` calculated in `performace_test` method - we take there `optimize_ctx`.

kfojcik-intel replied:
Hi @desertfire,
Could you take a look here?

kfojcik-intel committed:
Merge branch 'pytorch:main' into kfojcik/hpu_dynamo_benchmark

kfojcik-intel committed:
Merge branch 'pytorch:main' into kfojcik/hpu_dynamo_benchmark

kfojcik-intel committed:
Simplify  Gaudi support to benchmarks/dynamo

kfojcik-intel commented:
In the last commit [ec0d190](https://github.com/pytorch/pytorch/pull/139205/commits/ec0d1908aa86a359ce57eb4fc7605a7e99de726a)
I added some simplifications of this change:
- removed part where we switch the value of device param from "hpu" to "cuda". Instead, proposed a couple of conditions "`if device in ["cuda", "hpu"]`",
- removed `optimized_ctx` parameter from `speedup_experiment()` and connected functions since `optimized_ctx` is a global variable.

kfojcik-intel commented:
Hi @EikanWang @desertfire ,
Could you take a look here? :)

kfojcik-intel commented on benchmarks/dynamo/common.py:
```diff
@@ -3208,15 +3208,15 @@ def warmup(fn, model, example_inputs, mode, niters=10):
             start_stats = get_dynamo_stats()
             try:
-                if current_device == "cuda":
+                if current_device in ["cuda", "hpu"]:
                     torch.cuda.reset_peak_memory_stats()
```
We are using our GPU Migration Toolkit to handle it but we can also call our API directly.
I changed it in new PR: #145920 since I'm closing this one due to high complexity

kfojcik-intel commented:
Due to high complexity, I'm closing this PR. The simpler PRs are created: #145920 and #145921
